[PR2] DEV-23 improve connection negotiation and cleanup

jmaralo opened the pull request:
The previous implementation was out of sync with the media server. This new version uses the onnegotiationneeded event to know when to exchange offer and answer through the singaling channel.

Some events have also been buffered in case they are triggered during an unappropiate moment during the handshake.

The WebRTC connection is now correclty cleaned up after use, which caused errors previously with react strict mode due to incorrect cleanup.

The package has been modified to expose the development build on the network to make testing of the cameras page easier.

smorsua committed:
Fixed TubePage cameras

jmaralo committed:
DEV-23-2 improve connection negotiation

Sync signal specification with media server

Reliable negotiation with message buffering and use of webrtc events

Expose development version to the network with npm run dev

jmaralo committed:
Merge branch 'features/DEV-23-2' of https://github.com/HyperloopUPV-H8/control-front into features/DEV-23-2

Sync local changes with origin

jmaralo committed:
Merge branch 'develop' of https://github.com/HyperloopUPV-H8/control-front into features/DEV-23-2

smorsua commented on src/hooks/WebRTC/RemotePeer.ts:
```diff
@@ -0,0 +1,95 @@
+    signal: SignalChannel
+    private candidateBuffer = new Array<RTCIceCandidateInit>()
+
+    constructor(signalAddr: string, configuration?: RTCConfiguration) {
```
Remove configuration if not needed

smorsua commented on src/hooks/WebRTC/SignalChannel.ts:
```diff
@@ -1,74 +1,76 @@
         }
 
-        this.socket.onclose = () => this.socket.close
+        this.socket.onclose = () => {
```
The onclose and the onerror and not necessary

smorsua commented on src/hooks/WebRTC/signal.ts:
```diff
@@ -1,29 +1,14 @@
-};
-
-enum SignalCode {
+export enum SignalCode {
```
Make strings instead of numbers

smorsua commented on src/pages/CamerasPage/CamerasPage.tsx:
```diff
@@ -5,11 +5,7 @@ import { useWebRTC } from "hooks/WebRTC/useWebRTC";
-        { index: 1, url: import.meta.env.VITE_CAMERA_1_URL },
-        { index: 2, url: import.meta.env.VITE_CAMERA_1_URL },
-    ]);
+    const [streams, state] = useWebRTC(import.meta.env.VITE_CAMERA_1_URL);
```
Change env variable name to `VITE_CAMERAS_URL`

jmaralo committed:
DEV-23 remove unnecessary code

Remove SignalCode, ClosePayload and RejectPayload as it was deprecated

Remove useless onclose and onerror event listeners

Change env variable name

Losina24 commented on src/hooks/WebRTC/RemotePeer.ts:
```diff
@@ -0,0 +1,95 @@
+    }
+
+    private async onIceCandidate(ev: RTCPeerConnectionIceEvent) {
+        if (!ev.candidate) {
```
if (!ev.candidate) return

Losina24 commented on src/hooks/WebRTC/RemotePeer.ts:
```diff
@@ -0,0 +1,95 @@
+        this.close()
+    }
+
+    private async onCandidateSignal(signal: Signal<"candidate">) {
```
u have to try to remove this ifs with a lot of or clauses
```ts
private async onCandidateSignal (signal: Signal<"candidate">) {
    if (isCandidate()) return this.candidateBuffer.push(signal.payload)
    await this.peer.addIceCandidate(signal.payload)
}

private isCandidate() {
    const candidateStates = new Set(["stable", "have-remote-offer", "have-remote-pranswer"])
    return candidateStates.has(this.peer.signalingState)
}
```

Losina24 commented on src/hooks/WebRTC/RemotePeer.ts:
```diff
@@ -0,0 +1,95 @@
+    }
+
+    private async onSignalingStateChange() {
+        if (this.peer.signalingState !== "stable" && this.peer.signalingState !== "have-remote-offer" && this.peer.signalingState !== "have-remote-pranswer") {
```
same here

Losina24 replied:
to reuse the Set, u can declare outside the first function and use it in both functions

jmaralo committed:
DEV-23-2 use Set to check signaling state

Use a set to check if a candidate can be added to the peer
